llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1503,21 +1553,17 @@ std::pair<Type *, Value *> SPIRVToLLVM::createLaunderRowMajorMatrix(Type *const
                                                                     Value *const pointerToMatrix) {
   Type *const matrixPointerType = pointerToMatrix->getType();
 
-  assert(matrixType->isArrayTy() && matrixType->getArrayElementType()->isStructTy());
-
-  Type *const columnVectorType = matrixType->getArrayElementType()->getStructElementType(0);
-  assert(columnVectorType->isArrayTy());
-
-  // Now we need to launder the row major matrix type into a column major one.
-  Type *const newColumnVectorType =
-      ArrayType::get(columnVectorType->getArrayElementType(), matrixType->getArrayNumElements());
-  Type *const newMatrixType = ArrayType::get(newColumnVectorType, columnVectorType->getArrayNumElements());
+  Type *const newMatrixType = getTransposedType(matrixType);
   Type *const newMatrixPointerType = newMatrixType->getPointerTo(matrixPointerType->getPointerAddressSpace());
 
-  FunctionType *const rowMajorFuncType = FunctionType::get(newMatrixPointerType, matrixPointerType, false);
+  FunctionType *const rowMajorFuncType =
+      FunctionType::get(newMatrixPointerType, {matrixPointerType, matrixType}, false);
   Function *const rowMajorFunc =
       Function::Create(rowMajorFuncType, GlobalValue::ExternalLinkage, SpirvLaunderRowMajor, m_m);
-  return std::make_pair(newMatrixType, getBuilder()->CreateCall(rowMajorFunc, pointerToMatrix));
+  Value *dummyValue = Constant::getNullValue(matrixType);
+  Value *call = getBuilder()->CreateCall(rowMajorFunc, {pointerToMatrix, dummyValue});
```

Comment:
You'll have to add a type mangling suffix to the `rowMajorFunc` function name, to support cases where matrices with different element types and/or different sizes are used within a single shader.

Reply:
This is a very good point, thanks @nhaehnle . Now I'm trying to understand how this works right now, because we can have different `matrixPointerTypes` and we are not mangling the name (like **row_major.p7v3i64) and still this is working. When I was debugging it, I remember that LLVM was adding number as a suffix to the name.

Reply:
Ah, interesting point. I suppose we could rely on LLVM just appending a suffix.

Reply:
I added function name mangling as you suggested. This may be bit cleaner because right now we are creating multiple launderRowMajor (with different numeric suffix) functions even for the same input types.